### PR TITLE
feat: Enable Rule CA1815

### LINKED ISSUE #8945: [Performance] Override Equals on value types
**Is your feature request related to a problem? Please describe.**
https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1815

### Benchmark

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <OutputType>Exe</OutputType>
  </PropertyGroup>
  <PropertyGroup>
    <PlatformTarget>AnyCPU</PlatformTarget>
    <DebugType>pdbonly</DebugType>
    <DebugSymbols>true</DebugSymbols>
    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
    <Optimize>true</Optimize>
    <Configuration>Release</Configuration>
    <IsPackable>false</IsPackable>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="BenchmarkDotNet" Version="0.12.1" />
    <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="0.12.1"/>
  </ItemGroup>
</Project>
```

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System;

namespace testbench;

public partial struct FlowLayoutAnchorInfo
{
    internal FlowLayoutAnchorInfo(in int index, in double offset)
    {
        Index = index;
        Offset = offset;
    }
    public int Index;

    public double Offset;
}

public partial struct FlowLayoutAnchorInfo2
{
    internal FlowLayoutAnchorInfo2(in int index, in double offset)
    {
        Index = index;
        Offset = offset;
    }
    public int Index;

    public double Offset;

    public override bool Equals(object obj)
    {
        return obj is FlowLayoutAnchorInfo2 info &&
               Index == info.Index &&
               Offset == info.Offset;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Index, Offset);
    }

    public static bool operator ==(FlowLayoutAnchorInfo2 left, FlowLayoutAnchorInfo2 right)
    {
        return left.Equals(right);
    }

    public static bool operator !=(FlowLayoutAnchorInfo2 left, FlowLayoutAnchorInfo2 right)
    {
        return !(left == right);
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run(typeof(Program).Assembly);
    }
}


public class Benchmarks
{
    [Benchmark]
    public bool Scenario1()
    {
        var a = new FlowLayoutAnchorInfo(0, 0);
        var b = new FlowLayoutAnchorInfo(5, 0);
        return a.Equals(b);
    }

    [Benchmark]
    public bool Scenario2()
    {
        var a = new FlowLayoutAnchorInfo(0, 0);
        var b = new FlowLayoutAnchorInfo(0, 0);
        return a.Equals(b);
    }

    [Benchmark]
    public bool Scenario3()
    {
        var a = new FlowLayoutAnchorInfo2(0, 0);
        var b = new FlowLayoutAnchorInfo2(5, 0);
        return a.Equals(b);
    }

    [Benchmark]
    public bool Scenario4()
    {
        var a = new FlowLayoutAnchorInfo2(0, 0);
        var b = new FlowLayoutAnchorInfo2(0, 0);
        return a.Equals(b);
    }
}
```
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19044
Intel Core i5-10400 CPU 2.90GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=7.0.100-preview.2.22153.17
  [Host]     : .NET Core 6.0.6 (CoreCLR 6.0.622.26707, CoreFX 6.0.622.26707), X64 RyuJIT
  DefaultJob : .NET Core 6.0.6 (CoreCLR 6.0.622.26707, CoreFX 6.0.622.26707), X64 RyuJIT


```
|    Method |       Mean |     Error |    StdDev |     Median |
|---------- |-----------:|----------:|----------:|-----------:|
| Scenario1 | 145.666 ns | 2.6010 ns | 2.8910 ns | 145.938 ns |
| Scenario2 | 245.279 ns | 4.7885 ns | 5.7003 ns | 243.636 ns |
| Scenario3 |   3.230 ns | 0.0928 ns | 0.0992 ns |   3.222 ns |
| Scenario4 |   3.543 ns | 0.0999 ns | 0.2412 ns |   3.439 ns |

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by workgroupengineering**
## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes #8945
Partof #8944

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027079-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Media/BoxShadows.cs [hunk 1]
```diff
@@ -62,7 +62,7 @@ public override string ToString()
         }
 
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public struct BoxShadowsEnumerator
+        public record struct BoxShadowsEnumerator
```
Comment (maxkatz6) [inline src/Avalonia.Base/Media/BoxShadows.cs:65]: I don't think there is any point on declaring enumerators as records. Warning should be suppressed here and other enumerators IMO. They are not compared in any way.

File: src/Avalonia.Base/Animation/Cue.cs [hunk 2]
```diff
@@ -68,6 +68,11 @@ public bool Equals(double other)
         {
             return CueValue == other;
         }
+
+        public override int GetHashCode()
```
Comment (maxkatz6) [inline src/Avalonia.Base/Animation/Cue.cs:72]: Why do we need GetHashCode here if this struct is already a record?

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027165-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0027428-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-nightly/nuget/v3/index.json) [PRBUILDID]

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**